### ISSUE TITLE
Add support for task lists

### ISSUE THREAD

**Issue by ZachBacon** [s9e/TextFormatter#141]
so I've been using https://github.com/AlfredoRamos/phpbb-ext-markdown and the author mentioned that his extension leverages work from here and I was wondering if we can get markdown syntax extended to include things like check boxes etc.

**Comment by JoshyPHP** [comment]
No, but anybody can create plugins and their own markup.

https://s9etextformatter.readthedocs.io/Plugins/Preg/Synopsis/
https://s9etextformatter.readthedocs.io/Plugins/Your_own_plugin/Register_parser/

**Comment by JoshyPHP** [comment]
I'm reopening this issue with a new title. While I don't want to add task lists to the Litedown plugin because it's a fairly niche feature, I'm considering implementing it as a separate plugin.

What's your use case for task lists? How would you(r users) use them?

**Comment by JoshyPHP** [comment]
@AlfredoRamos Current WIP branch is in https://github.com/s9e/TextFormatter/pull/142

I have no familiarity with task lists so I'm looking for use cases and opinions on how it should work. AFAIK GitHub pioneered this markup and called it "task lists" but in reality they're *checklists*. I would expect a *task* list to have more than 2 states, maybe something like "in progress", "done", "abandoned" but AFAICT all of the implementations I found (GitHub / GitLab / MD4C) only support checkboxes.

**Comment by JoshyPHP** [comment]
About the API: I think it only makes sense to have task lists if they can be toggled directly from the document. For that purpose, each checkbox gets a unique `data-task-id` value that is randomly generated during parsing. I'm thinking about having an API like `checkTask(string $xml, string $id): string` (and its counterpart `uncheckTask`) to check/uncheck a specific task and get the new XML in return. I'll leave the UI and controller to implementers.

By default, the checkboxes are disabled with an option to make them editable at rendering time.

**Comment by AlfredoRamos** [comment]
Yes, usually they are used like a to-do list, where there's only two states:

- Complete (checked) the task has been completed
- Incomplete (unchecked), the task for some reason has not been done yet, or it was cancelled

Example:

- [x] Play with my dogs
- [ ] Call my brother
- [ ] ~Go to the gym~

As far as I know a `checkbox` can have up to 3 states:

- Checked
- Unchecked
- Indeterminate

However this last one only works setting that property though JavaScript ([reference]).

[reference]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#Indeterminate_state_checkboxes

**Comment by JoshyPHP** [comment]
I think I'm just about done with the plugin. I settled for "complete" and "incomplete" to describe the state of a task, mostly because that's what GitHub uses to describe tasks in their docs.

@AlfredoRamos Check out the new docs, let me know what you think: https://github.com/s9e/TextFormatter/blob/2.5/TaskLists/docs/Plugins/TaskLists/Synopsis.md

**Comment by AlfredoRamos** [comment]
@JoshyPHP it works perfectly! I've updated my pull request https://github.com/AlfredoRamos/phpbb-ext-markdown/pull/43

It even works better that the GitHub implementation, because my test markup is not rendered well here, and I'm almost sure it's a valid markup.

I don't know what's the "standard" for tasks list (if any exists at all), but in my humble opinion, a task item without text should not be rendered, as they do not provide useful information of what they are:

![TaskList plugin](https://i.imgur.com/yiqRFaH.png)

<details>
<summary>Test markup</summary>

```
Task lists:

1. [x] Done
2. [ ] To do


* [x]
* [x]


- [x] Task 1
  - [x] Task 1.1
    - [x] Task 1.1.1
  - [x] Task 1.2
- [ ]
- [x]
- [ ] Task 2
  - [ ] Task 2.1
  \```shell
find . -type f -exec chmod 644 {} \; && find . -type d -exec chmod 755 {} \;
  \```
    - [ ] Task 2.1.1
      - [x] Task 2.1.1.1
- [ ] Task 3
```

**Note:** I had to escape the code block in task 2.1
</details>

**Comment by JoshyPHP** [comment]
Quick update. I intend to squash-merge this branch in the upcoming 2.5.0 release, which I optimistically expect to publish next week.

> I don't know what's the "standard" for tasks list (if any exists at all), but in my humble opinion, a task item without text should not be rendered, as they do not provide useful information of what they are

I've thought about it and there doesn't seem to be a consensus about whether a task must be followed by something. It varies between implementations and I think each implementation went with whatever was simpler for them at the time. I looked at [how Markdown libraries handle empty list items](https://johnmacfarlane.net/babelmark2/?normalize=1&text=*+1%0A*%0A*+3) and the most recent implementations seem to follow the CommonMark flavour that intentionally allows them whereas the original Markdown would treat the empty list item as a continuation of the one before. So following this trend, I think it's better to allow empty tasks. It's also simpler to implement.

**Comment by JoshyPHP** [comment]
Implemented in 2.5.0.
